### PR TITLE
Add multilingual

### PR DESCRIPTION
**Pull request by marktopper**
# Progress:
- [x] Add `translations` migration
- [x] Add `Translatable` trait
- [x] Add method to get translated value
- [x] Add method to translate the whole model
- [x] Enable saving directly to the translator
- [x] Enable checking if a model is translatable
- [x] Create view to input translated values (#873)
- [x] Make method to get original model from translator
- [x] Make method to create translation
- [x] Make method to update translation
- [x] Make method to delete translation
- [x] Make method to get meta data for translation
- [x] ~Make method to get rows that have translations~~ _(can come later as a PR)_
- [x] ~Make method to get rows that does not have translations~~ _(can come later as a PR)_
- [x] Make tests
- [x] ~Turn this into a plugin for Voyager~ _(Cancelled)_

# Setup:

In order to use this multilingual, you must set some extra languages in the configuration. It will not work where there are only the default language.

Then you must ensure that the models you wish to use the multilingual support for have the `TCG\Voyager\Translatable` trait.
Along with that you must define the translatable attributes in the model as so
```
class Page extends Model
{
    protected $translatable = ['title', 'body'];
}
```

# Usage:

## Eager load translations
```
// Loads all translations
$posts = Post::with('translations')->get();

// Loads all translations
$posts = Post::all();
$posts->load('translations');

// Loads all translations
$posts = Post::withTranslations()->get();

// Loads specific locales translations
$posts = Post::withTranslations(['en', 'da'])->get();

// Loads specific locale translations
$posts = Post::withTranslation('da')->get();

// Loads current locale translations
$posts = Post::withTranslation('da')->get();
```

## Get default language value
```
echo $post->title;
```

## Get translated value
```
echo $post->getTranslatedAttribute('title', 'locale', 'fallbackLocale');
```
> If you do not define `locale`, the current application locale will be used. You can pass in your own locale as a string.
> If you do not define `fallbackLocale`, the current application fallback locale will be used. You can pass in your own locale as a string. If you want to turn the fallback locale off, pass `false`.
> If no values are found, `null` is returned.

## Translate the whole model
```
$post = $post->translate('locale', 'fallbackLocale');
echo $post->title;
echo $post->body;

// You can also run the `translate` method on the Eloquent collection
// to translate all models in the collection.
$posts = $posts->translate('locale', 'fallbackLocale');
echo $posts[0]->title;
```
> If you do not define `locale`, the current application locale will be used. You can pass in your own locale as a string.
> If you do not define `fallbackLocale`, the current application fallback locale will be used. You can pass in your own locale as a string. If you want to turn the fallback locale off, pass `false`.
> If no values are found for the model for a specific attribute, either for the locale or the fallback, it will set that attribute to `null`.

## Check if model is translatable
```
// with string
if (Voyager::translatable(Post::class)) {
    // it's translatable
}

// with object of Model or Collection
if (Voyager::translatable($post)) {
    // it's translatable
}
```

## Save changes
```
$post = $post->translate('da');
$post->title = 'foobar';
$post->save();
```
> This will update or create the translation for `title` the post with the locale `da`.
> Please note that if a modified attribute is not translatable, then it will make the changes directly to the model itself. Meaning that it will overwrite the attribute in the language set as default.

# Feedback, questions and pull requests are more than welcome.

### CODE REVIEW

**Comment by marktopper** [comment]
__UPDATE__: Added new ways to eager load translations.

**Comment by marktopper** [comment]
__UPDATE__: Added new ways to eager load translations.

**Comment by marktopper** [comment]
__UPDATE__: You can now translate a Eloquent Collection like this.

```
$posts = Post::all()->translate('locale');
echo $posts[0]->title; // is now translated
```

You can also make modifications to the translated models and save them as translations like this.
```
$post = Post::first()->translate('locale');
$post->title = 'test';
$post->save();
```
> Please note that if a modified attribute is not translatable, then it will make the changes directly to the model itself. Meaning that it will overwrite the attribute in the language set as default.

**Comment by marktopper** [comment]
__UPDATE__: You can now use the Voyager facade to check if a object or classname is translatable.

```
// with string
if (Voyager::translatable(Post::class)) {
    // it's translatable
}

// with object of Model or Collection
if (Voyager::translatable($post)) {
    // it's translatable
}
```

File: src/Translator/Collection.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace TCG\Voyager\Translator;
+
+use Illuminate\Support\Collection as IlluminateCollection;
+
+class Collection extends IlluminateCollection
+{
+    //
```
Comment (fletch3555) [inline src/Translator/Collection.php:9]: @marktopper, what's the purpose of this file? It has no contents, and doesn't override anything from the `IlluminateCollection`

**Comment by akazorg** [comment]
Hello @marktopper!
This is a very interesting subject, since we can handle it in so different ways.

I was curious for your approach, I see you opted for the extra table, in fact we can gain a lot of features like user/author control, history, etc. But do we really need this?

I'd rater see the JSON approach, with the spatie/translatable. I have a working version of that, but need some time to create a fork. We can work on something together if you want.

Some screen shoots (you can also see the slugify in action):

**BREAD**
peek_bread_2017-01-15 02-35

**Menu**
peek_menu_2017-01-15 02-39

File: src/Translator/Collection.php (hunk 1 above)
Comment (marktopper) [reply src/Translator/Collection.php:9]: Custom macros can be added only to the collection of the translators.
And will add some custom helper methods there later.

**Comment by marktopper** [comment]
Hey @akazorg.

Thanks a lot for the screenshots, this seems like the best approach when it comes to the UI.
Will look into implementing this.
Sure, lets work on something together, that is also one of the main reasons that I have put it up as a branch that everyone can make pull requests to.
About the extra table, I just wanted to let it be flexible. However I can make it extendable so that people can easily make it into JSON instead and remove the `translations` table.

**Comment by akazorg** [comment]
Hello @marktopper, 

I will create a fork so you can access that code.
Meanwhile I'll test your version, and see how the UI could be implemented. I've wrote a JS, check out [this guist](https://gist.github.com/akazorg/32537234488b6397e864a63b951ba2ff), some notes included.

**Comment by marktopper** [comment]
That is a quite nice script you have made there @akazorg.

Thanks for sharing. And let me know how it goes for the UI.
I will meanwhile work on some more helpers for it.
And maybe do some examples on usage at the frontend.

**Comment by akazorg** [comment]
Thanks @marktopper.
That JS is working, just requires some revision because the initial version was aimed to work has a BREAD option, and I ended up changing the design and defining translations at Models, like you did.

It would be awesome to let user define translations has a BREAD option, this way they didn't have to edit Models, but there's some work involved. Anyway, just sharing an idea.

Happy coding, bbl with news

**Comment by marktopper** [comment]
I like your idea @akazorg.

Well, there are two solutions for that.

1.- Editing the model when setting translatable inputs in the UI.
2.- Save the translatable fields somewhere else.

I however love the idea to make it work from the UI instead of the current way where they have to enter it directly in the model.

**Comment by marktopper** [comment]
For updating a model's attributes, we can do this the following way:

```
$attribute = 'translatable'; 

$reflector = new ReflectionClass(\App\Page::class);
$path = $reflector->getFileName();

$content = file_get_contents($reflector->getFileName());

if (preg_match('/protected \$'.$attribute.' = (?:\[|array\()([\n\t\r\w\d\s\',]*?)+?(\]|\));/im', $content)) {
	$content = preg_replace('/protected \$guarded = (?:\[|array\()([\n\t\r\w\d\s\',]*?)+?(\]|\));/im', 'foo', $content);
} else {
	$content =preg_replace('/(class ([\n\t\r\w\d\s\':,]*?)+(\n*)?{)/im', "class $1\n    protected ${$attribute} = [];", $content);
}

file_put_contents($reflector->getFileName(), $content);
```

Or we could set an option to use some values from the database instead. Still open for ideas on this one.

**Comment by akazorg** [comment]
The idea of the option is great, seems to be more clear.

**Comment by marktopper** [comment]
Just created 12 new tests to handle some of the core and basic features for the multilingual system.

**Comment by marktopper** [comment]
@akazorg: My plan is to add the multilingual feature now and then later make it so that translatable attributes can be read from the database by setting `protected $translatable = null;` or something.

**Comment by marcoocram** [comment]
@marktopper it will be compatible anyway with https://github.com/dimsav/laravel-translatable?

Thanks!

**Comment by marktopper** [comment]
@marcoocram, that package requires a bit more to integrate, but later my plan is to make it as a plugin for Voyager and remove it from the core of Voyager (maybe by `v1.0` release). Then I will also use some time to do some alternative integrations like [that](https://github.com/dimsav/laravel-translatable).

**Comment by marktopper** [comment]
@akazorg: All I need know is the integration into the views. Can you share what you had for this so I do not have to reinvent the wheel?

**Comment by marktopper** [comment]
__UPDATE:__ I will try to turn this into a plugin now. Since there is not reason to have it to the core of Voyager.

**Comment by marcoocram** [comment]
I prefeer, what is the ETA?

**Comment by marktopper** [comment]
My ETA for publishing this plugin is one week after the release of Voyager [`v0.11`](https://github.com/the-control-group/voyager/milestone/3).

**Comment by marktopper** [comment]
And Voyager v0.11 should be done this sunday.

File: src/Traits/Translatable.php [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    /**
+     * Load translations relation.
+     *
+     * @return mixed
+     */
+    public function translations()
+    {
+        return $this->hasMany(Translation::class, 'foreign_key', $this->getKeyName())
+            ->where('table_name', $this->getTable())
+            ->whereIn('locale', config('voyager.multilingual.locales', []));
```
Comment (morloderex) [inline src/Traits/Translatable.php:35]: Maybe include a fallback to someting like `en` just incase :)

Comment (marktopper) [reply src/Traits/Translatable.php:35]: The default language is `en` and is not stored in `translations` but in the model itself.

Let's say a model `Post` have been there for quite a long time, and all it's content is in `en`, then the default language should be set to `en` in the configuration and then they can be translated into other languages.
The other languages will however be stored inside the `translations` table.
So if the user was ever to disable multilingual on the site, they would still have the original value of their posts in their main language.

**Comment by akazorg** [comment]
Hello @marktopper, sorry but I've been so busy with another project (not using Laravel), so I'm bit away.
This week I'll find time for working on this.

**Comment by marktopper** [comment]
@akazorg: It is totally okay. You can also just share you view and JS with me and then I can merge it into this.

However, no hurry. Gonna build this into a separate plugin instead for Voyager.

**Comment by Frondor** [comment]
This doesn't translates voyager's UI right? Can we use default L5.4 for that?

`resources/lang/en/voyager.php`

`__('voyager.readmore')`

As you know, lang files can be overriden like `resources/lang/vendor/voyager/en/voyager.php`

If that is not addressed yet, I can make a PR with both `EN` and `ES` files.

**Comment by marktopper** [comment]
No this only makes the BREAD translatable.

**Comment by ronrun** [comment]
Hi, how to use this multilingual without voyager? I don't need the permission tables, admin views. Is there installation guide?

**Comment by akazorg** [comment]
Hi @ronrun. For that (without using Voyager), I suggest using another multilingual library. 

Some options:
- [Spatie / laravel-translatable](https://github.com/spatie/laravel-translatable)
- [Barryvdh / laravel-translation-manager](https://github.com/barryvdh/laravel-translation-manager)
- [Dimsav / laravel-translatable](https://github.com/dimsav/laravel-translatable)

**Comment by mustafa-ramax** [comment]
Hello guys, 

I'm trying to call a post with another language in home.blade.php and i"m getting the error 
Method getTranslatedAttribute does not exist.

**Comment by akazorg** [comment]
Hi @Steve-sy, have you setup the config `voyager.php` and followed the instructions from docs?
If you can't sort that out, please open a new issue and we will help you.
Thanks.

**Comment by coppee** [comment]
Hello there,
Just to be clear, the final translation system choice is to use a specific table, no the json solution ?
And btw, thanks for all your great work !

**Comment by marktopper** [comment]
@coppee We use the database for the BREAD translations but will use PHP files for the views-translations.

**Comment by coppee** [comment]
Hi @marktopper , thanks for your reply. Sorry, I was not clear enough in my question. I was talking about the other translation eloquent solution (use json in fields) proposed by @akazorg like : https://github.com/spatie/laravel-translatable

**Comment by akazorg** [comment]
Hi @coppee Voyager is using the extra table approach, not the JSON one.
It works great this way.

**Comment by mbdwey** [comment]
Hello @marktopper
Am using your package to generate admin panel for mobile API
and I will translate the admin panel to Arabic please share more information about contribution guidelines.
I could help merging i18n and feature/rtl-support and test or improve them

**Comment by vitorbrussi** [comment]
Hello, I am new not Voyager, not understood how to translate to pt_br.
They could pass me one step at a time. Thank you.

**Comment by akazorg** [comment]
Hi @vitorbrussi, check this #1364.

**Comment by vitormicillo** [comment]
It would be very interesting if there was the possibility of having a auto-translator when a locale is defined, and return the translated data In fields that have translation

**Comment by akazorg** [comment]
That's a nice idea @vitormicillo, we should work on that.

**Comment by vitormicillo** [comment]
I created a not very elegant method but it works to auto translate the specific words
```
private function translatedCollectionToArray(\TCG\Voyager\Translator\Collection $translatedCollection) {
        $dateFields = [ 'created_at', 'updated_at', 'deleted_at' ];
        $collectionArray = array();
        foreach($translatedCollection as $collectionElement)
        {
            $collectionArray[] = array_map(function($rawAttributeData) { return $rawAttributeData['value']; }, $collectionElement->getRawAttributes());
        }
        foreach($collectionArray as $i => $collectionElement)
        {
            foreach($collectionElement as $elementKey => $elementValue)
            {
                if(!in_array($elementKey, $dateFields) || $elementValue == null) { continue; }
                $collectionArray[$i][$elementKey] = strtotime($elementValue);
            }
        }
        return $collectionArray;
    }
```

**Comment by vitormicillo** [comment]
This method recycles the use of the '\TCG\Voyager\Translator\Collection'
And stamps the dates in timestamp
I hope to help a little, to run it just send in the header your locale
and to call this methodo:
I'm using: 'Sections' => $this->translatedCollectionToArray(Section::all()->translate($language, 'fallbackLocale')),

$language = $request->header('locale');
Section::all() You can modify to Post::all() to try or Category::all()

**Comment by DODMax** [comment]
Really great work on this multilingual feature!

I'm trying to think of a way to use it to store my Laravel translations (like in https://github.com/Waavi/translation but I'd like to avoid using a different package/ UI)

So basically I'm with my translation files that look something like this
```
'date' => 'Payment date: :date',
'amount' => 'Amount: :amount :currency',
'plan' => 'Plan: :plan',
'from' => 'From: :date',
'to' => 'To: :date',
```

And I'm thinking of adding a `labels` table with my `key` and `label` columns, making it translatable then having a class a bit like this:
```php
class VoyagerTranslator extends Illuminate\Translation\Translator
{
	private $labels = null;

	public function get($key, array $replace = [], $locale = null, $fallback = true)
	{
		$lang = $this->getLabel($key);

		if ($lang && ($line = $lang->getTranslatedAttribute('label', $locale, false))) {
			return $this->makeReplacements($line, $replace);
		} else {
			return parent::get($key, $replace, $locale, $fallback);
		}
	}

	protected function getLabel($key)
	{
		if ($this->labels == null) {
			$this->labels = \App\Label::with('translations')->get();
		}

		return $this->labels->where('key', $key)->first();
	}
}
```

Any thoughts? Thanks!

EDIT:  Ok after some edits this seems to work well: this will use Voyager's translation from DB if present otherwise fall back to Laravel language files (once registered with the proper service provider of course).

**Comment by vitormicillo** [comment]
I created a method to use the translations

    private function translatedCollectionToArray(\TCG\Voyager\Translator\Collection $translatedCollection) {
        $collectionArray = array();
        foreach($translatedCollection as $collectionElement)
        {
            $collectionArray[] = array_map(function($rawAttributeData) { return $rawAttributeData['value']; }, $collectionElement->getRawAttributes());
        }
        return $collectionArray;
    }

and to use:

Ex.

$language = $request->header('locale');
if($language == null || $language !== "it"){$language = "en";}
$section = $this->translatedCollectionToArray(Posts::all()->translate($language, 'fallbackLocale'));

works fine for me to use in translating api, I hope it helps many

**Comment by sergbbb** [comment]
Please include this guide in Documentation. It is hard to find it. Thank you.

**Comment by yarokov** [comment]
Hi! How can I hide the not translated posts?

**Comment by fadighattas100** [comment]
Hi, any video tutorial for how to do this like in the gif's images?  how to translate the post's content , thanks in advance.

**Comment by PointfullDevelopers** [comment]
I tried to implement this, but no luck.

1. When I add ['en' => 'English', 'sv' => 'Swedish'], to the 'locale' on the config it gives error.
2. There is no Model for Posts and Pages so, where do I make those changes?

Please update your documentation

**Comment by amrittb** [comment]
It seems that I can't translate the entire model and other loaded models using Eloquent relationships using the above methods. Any chance of support on this?

**Comment by emptynick** [comment]
Certainly. But you have to give more informations.
And please ask questions in our Slack group!

**Comment by sten** [comment]
Hiu @marktopper, you mentioned about a year ago that you wanted to make the translation library in voyager configurable: "About the extra table, I just wanted to let it be flexible. However I can make it extendable so that people can easily make it into JSON instead and remove the translations table."

Is the current implementation configurable? Could you point me to how you would replace the current translation implementation with for instance the one from Spatie: [spatie/laravel-translatable](https://github.com/spatie/laravel-translatable)

Thanks!

**Comment by AlexanderWM** [comment]
@marktopper Any ideas how to fix #3089 ?

**Comment by fletch3555** [comment]
@AlexanderWM, this PR has been merged. 3089 would be a much better place to ask that question

**Pull request merged**